sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -64,10 +64,17 @@ object PhysicalOperation extends PredicateHelper {
         val substitutedFields = fields.map(substitute(aliases)).asInstanceOf[Seq[NamedExpression]]
         (Some(substitutedFields), filters, other, collectAliases(substitutedFields))
 
-      case Filter(condition, child) if condition.deterministic =>
+      case Filter(condition, child) =>
         val (fields, filters, other, aliases) = collectProjectsAndFilters(child)
-        val substitutedCondition = substitute(aliases)(condition)
-        (fields, filters ++ splitConjunctivePredicates(substitutedCondition), other, aliases)
+
+        // Deterministic parts of filter condition placed before non-deterministic predicates could
+        // be pushed down safely.
+        val (pushDown, rest) =
```

Comment:
@cloud-fan Now that I could insert a `Scanner` operator over `CatalogRelation` in `Optimizer`, but I noticed a relation may also be something like `l: LogicalRelation(relation: CatalogRelation, _, _)`, in this case, we couldn't analyze the class `LogicalRelation` because it's in package `spark-sql` while `Optimizer` is in `spark-catalyst`, thus we are not able to determine whether a `Scanner` should be added. I think we don't want to add `Scanner` over every `BaseRelation`.